### PR TITLE
[CST-4009] Discovery result sort options not reflecting what is configured

### PR DESCRIPTION
**Pull request by alemarte**
## References
* Fixes https://github.com/DSpace/dspace-angular/issues/1071

## Description
As described in the related issue, sort options aren't provided correctly to the SearchSettingCompontent, which is only provided with fixed default values.

It has been added a SearchConfig model and a method to fetch the search configuration given scopes and configuration names.
Then this information is propagated by the root component (MyDspacePageComponent) to the child components.

## Instructions for Reviewers
* Navigate to the [MyDspace](http://localhost:4000/mydspace?configuration=workspace) and check the available sort options are correct.
* Should be tested with https://github.com/DSpace/DSpace/pull/3241

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Comment by alemarte** [comment]
This is not a super clean solution for addressing the ticket, but the only one affordable in terms of time to fix this in beta5. This is due to the previous way to retrieve defaults and to update search params subsequently to route changes. In particular almost all the logic to fetch and initialize the sort options has been bounded in the MyDspacePageComponent initialization, but it is clear that should belong to the SearchConfigurationService and every component should manage its initialization taking this async into account (impact  in many components and their test suites).

- The defaults are computed synchronously and used by every component which is provided by its search configuration service. The logic should be asynchronous and so every components needs to change the way they initialize themselves. This should be an overall refactoring and needs a separate ticket. 
- This can be implemented in the context of fixing another big design problem of the SearchConfigurationService which therefore already needs some design refactoring. The problem is the moltiplication of search rest call when more than one queryparam changes after filter/facet/sort selection. A way to test is to change the sort options from score/ASC to title/DESC, two request are performed (title/ASC and then title/DESC).

Also, since the rest endpoints provide as sortOptions just the list of the fields, the generation of the available options now happens as the following (open to discussion):

- Rest order give the displayed priority.
- Every field has by default both direction available, ASC and DESC (first displayed ASC).
So for example:
`sortOptions: [{name: "dc.title"}, {name: "dc.date.issued"}, {name: "score"}]`
lead to options like this:
`dc.title.ASC, dc.title.DESC, dc.date.issued.ASC, dc.date.issued.DESC, score.ASC, score.DESC`

With this approach, if there was the necessity to exclude an option (for instance score ASC), it would be filtered directly angular side with a deny list.

To conclude, the Rest Contract of the discover/search is not accurated and need to be allineated to the actual rest implementation. Since we now follow the rest order, we noticed also that the default configured option (in our case score) is appended at the end instead of being inserted at the beginning of the list. This is something that should be fixed as well.

**Comment by artlowel** [comment]
Thanks @alemarte!

>This is not a super clean solution for addressing the ticket, but the only one affordable in terms of time to fix this in beta5. >This is due to the previous way to retrieve defaults and to update search params subsequently to route changes. In particular almost all the logic to fetch and initialize the sort options has been bounded in the MyDspacePageComponent initialization, but it is clear that should belong to the SearchConfigurationService and every component should manage its initialization taking this async into account (impact in many components and their test suites).
>
>- The defaults are computed synchronously and used by every component which is provided by its search configuration service. The logic should be asynchronous and so components need to change the way they initialize themselves. This should be an overall refactoring and needs a separate ticket.

With this PR the sort options only work in mydspace though, was that your point? The regular search page [also has sortOptions](https://api7.dspace.org/server/#https://api7.dspace.org/server/api/discover/search), yet none are shown with this PR. 

>The problem is the moltiplication of search rest call when more than one queryparam changes after filter/facet/sort selection. A way to test is to change the sort option from score/ASC to title/DESC, two requests are performed (title/ASC and then title/DESC). See subscribeToSearchOptions and subscribeToPaginatedSearchOptions

I struggled to reproduce this at first: the thing is that Relevance (i.e. score) is already sorted DESC, so you should switch from Relevance to Title ascending to reproduce it. I can confirm that this also happens on the demo site, and is therefore not caused by this PR.

>Also, since the rest endpoints provide as sortOptions just the list of the fields, the generation of the available options now happens as the following (open to discussion):
>
>- Rest order gives the displayed priority.
>- Every field has by default both direction available, ASC and DESC (first displayed ASC).
>So for example:
>`sortOptions: [{name: "dc.title"}, {name: "dc.date.issued"}, {name: "score"}]`
>leads to options like this:
>`dc.title.ASC, dc.title.DESC, dc.date.issued.ASC, dc.date.issued.DESC, score.ASC, score.DESC`
>
>With this approach, if there was the necessity to exclude an option (for instance score ASC), it would be filtered directly angular side with a deny list.

That will do for now, but in my opinion rest should also provide the available sort directions eventually.

>To conclude, the Rest Contract of the discover/search is not accurated and need to be allineated to the actual rest implementation. Since we now follow the rest order, we noticed also that the default configured option (in our case score) is appended at the end instead of being inserted at the beginning of the list. This is something that should be fixed as well.

Is that something that can be fixed with configuration on the rest side? A better default config might be all we need here?

**Comment by atarix83** [comment]
> With this PR the sort options only work in mydspace though, was that your point? The regular search page also has sortOptions, yet none are shown with this PR.

@artlowel we've added the code to retrieve sortoptions also for the regular search page

**Review by artlowel — CHANGES_REQUESTED** [review]
Thanks @atarix83.

The code looks good with the caveat you mentioned, that the init code should eventually move to SearchConfigurationService

Some feedback on the way it works though:
- If you change the sort option a number of times in a row the option in the URL will always be accurate, however the option shown in the select, and the param used in the requests to the backend will quickly lag behind one change. E.g. if I change it to Relevance desc, then to Date issued asc, then to Date accessioned desc, the url will correctly show `sortDirection=DESC&sortField=dc.date.accessioned` but the select will show 'Date issued asc', and the last rest request will have been to `objects?sort=dc.date.issued,ASC`
- The default sort option for the search page should be relevance descending. It's what everyone will expect. If we can fix it in config on the Rest side before the testathon, that's fine, otherwise I'd hardcode it to be the default on the client in this PR.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@alemarte and @atarix83 : I tested this today as well. I'm able to reproduce the same issues that @artlowel mentions.

Additionally, the behavior of the default search is odd, as the sort claims to be by title but seems to be by relevance.  Here's what I see
1. In the header, click the search icon. Type in "test"
2. You'll see the search results. The "Sort By" says "Title Ascending" (though for the results I see they do not seem ordered by title)
3. Click on page 2. The "Sort By" changes to "Relevance Ascending".  But, in the address bar of my browser it says: `&sortDirection=ASC&sortField=dc.title`
4. Click Reload in the browser, Now your page of results changes, as it reorders by "Title Ascending"

It looks, to me, like the "Sort By" field is not working properly or caching values at some times. This behaviour is also similar to the issue @artlowel reported (and I can reproduce), that if you change the sort several times, the "Sort By" value is not always correct.

Finally, in the "Sort By" dropdown I'm seeing some missing i18n keys for `sorting.dc.date.accessioned.ASC` and `sorting.dc.date.accessioned.DESC`

**Comment by alemarte** [comment]
@tdonohue @artlowel thanks for your feedback!

> @alemarte and @atarix83 : I tested this today as well. I'm able to reproduce the same issues that @artlowel mentions.
> 
> Additionally, the behavior of the default search is odd, as the "Sort By" says it's sorting by "Title Ascending" but the results themselves seem to be sorted by relevance. Here's what I see:
> 
> 1. (I'm logged out) In the header, click the search icon. Type in "test"
> 2. You'll see the search results. The "Sort By" says "Title Ascending" (though for the results I see they do not seem ordered by title, they look sorted by relevance to me.)
> 3. Click on page 2. The "Sort By" changes to "Relevance Ascending".  But, in the address bar of my browser it now says: `&sortDirection=ASC&sortField=dc.title`
> 4. Click Reload in the browser, Now your page of results changes, as it reorders results by "Title Ascending"
> 

I've committed a change that add sort field and sort direction to the active route at the initialization of SearchComponent and MyDspacePageComponent. With this change every request seems correct (even those involving a page change).
(During this tiny refactor I've also moved the initialization logic to the service, in this way we don't have so much code duplication).

> It looks, to me, like the "Sort By" field is not working properly or caching values at some times. This behaviour is also similar to the issue @artlowel reported (and I can reproduce), that if you change the sort several times, the "Sort By" value is not always correct.
> 

I noticed this but forgot since I've worked almost immediately upon https://github.com/DSpace/dspace-angular/pull/1048
and that PR seems to fix this. In fact after merging with main I wasn't able to reproduce it.

> Finally, in the "Sort By" dropdown I'm seeing some missing i18n keys for `sorting.dc.date.accessioned.ASC` and `sorting.dc.date.accessioned.DESC`

[Done](https://github.com/DSpace/dspace-angular/blob/d54b7d9f7c8cbcffea2c9165e35861d4b7d2a1c1/src/assets/i18n/en.json5#L3048)

**Comment by tdonohue** [comment]
Based on discussion in today's meeting, we've moved this ticket to the "7.0" release board. This means this work is still under review but is unlikely to be in Beta 5...mainly because the user experience with this PR isn't the greatest until we find a way to solve the REST API bugs described in https://github.com/DSpace/DSpace/issues/3227.  Therefore, we don't want to deploy this to the Testathon servers until the REST API bug is also solved (but as soon as both are ready, they will be deployed together).

File: src/assets/i18n/en.json5
Comment (tdonohue) [inline src/assets/i18n/en.json5:3046]: I'd recommend we change the text of these to be:

`"sorting.score.ASC": "Least Relevant"`
`"sorting.score.DESC": "Most Relevant"`

For me, at least, it's difficult to determine what "Relevance Descending" means....so, it may just be easier to work them in a more understandable way.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @alemarte !  This is now working for me, when I test it with the Backend PR: https://github.com/DSpace/DSpace/pull/3241

That said, I have a few tiny fixes to suggest.

First, the search results heading is missing a translation. It looks like it changed from the key `search.results.head` to `default.search.results.head`.  So, I think we need to rename this key: https://github.com/DSpace/dspace-angular/blob/bdc2dd5f9ca98ddb0e89371e02201cdb5ef1d94a/src/assets/i18n/en.json5#L3002

Second, I have another minor suggestion inline below.  Overall this looks great though & I cannot reproduce any of the previous errors.

File: src/assets/i18n/en.json5
Comment (alemarte) [reply src/assets/i18n/en.json5]: @tdonohue I've updated the suggested translations. I've kept a translation also for `search.results.head` since according to https://github.com/DSpace/dspace-angular/blob/8cade4fbf5635ff7fc2cce6e126ba008ebf30692/src/app/shared/search/search-results/search-results.component.html#L1

that label cannot be excluded a priori

**Review by tdonohue — APPROVED** [review]
👍 Thanks @alemarte ! This looks good to me now

File: src/app/core/shared/search/search-configuration.service.ts
Comment (artlowel) [inline src/app/core/shared/search/search-configuration.service.ts:258]: You could add the direction here that is specified in `sortOption` instead of both

**Review by artlowel — APPROVED** [review]
Thanks @alemarte! It works well.

I know we agreed earlier that for this PR you could simply add both sort directions because rest didn't provide the available ones. But it seems that with DSpace/DSpace#3241 rest _does_ provide the direction, and it looks to be easy to address on the UI side, so it would be nice if you could fix that in this PR as well

**Comment by alemarte** [comment]
@tdonohue @artlowel correct me if I'm wrong, but the server with DSpace/DSpace#3241 only provides the default direction. So we must assume that both are always available. This is something that maybe @abollini can easily confirm. 
Also we made this decision to keep the alternative directions with the pattern ASC/DESC because when you receive mixed defaults from the server the select list looks confusing and messy. So summing all up, you receive from the server something similar to this:

```
sortOptions: [
  {name: "score", sortOrder: "DESC" },
  {name: "dc.title", sortOrder: "ASC" },
  {name: "dc.date.issued", sortOrder: "ASC"}
]

```
and the UI displays the selection like this (bold is the selected option)

> score ASC
> **score DESC**
> dc.title ASC
> dc.title DESC
> dc.date.issued ASC
> dc.date.issued DESC

**Comment by artlowel** [comment]
A default direction per sort option has very limited use in my opinion. You could say it's relevant to configure that for score the default should be descending, but then again score ascending is completely useless and we should have a way to leave it out. The whole reason I noticed it was because now we have the sort option "Least relevant" on the search page. You'd have to come up with a _very_ creative query to make use of that 😄 

But if it was indeed the goal of the rest api to add a default direction per field, rather than to add all possible field and direction combinations separately, then this is something we should discuss on the rest side, and I'd just merge this PR as it is now and address it in a later PR. So it would be good if @abollini could weigh in here to clarify the intention.

**Comment by tdonohue** [comment]
@abollini : Could you weigh in here & verify what I'm seeing? @artlowel brings up some good points...but, it also seems like (to me) it might be a misunderstanding of what the REST API changes were meant to achieve?

Here's what I see:

1. The REST API PR (https://github.com/DSpace/DSpace/pull/3241) ensures that sort options come back as configured. Each `sortOption` now has its own `sortOrder` (which is set to the configured `defaultSortOrder` for that field as defined in `discovery.xml`).  This was redesigned from `discovery.xml` having a single, global "defaultSortOrder".
2. In this Angular PR, it appears it appears we are doing this:
    * Determining the first `sortOption` returned and using it's `sortOrder` to determine the default sort of the results. 
    * Listing all other `sortOption`s (ignoring their `sortOrder`)... so, it appears the only `sortOrder` that matters to the UI is the one set on the *first* `sortOption`

It appears to me that we've essentially just "redefined" how the global "defaultSortOrder" is/was configured/returned.  It used to be a separate setting in `discovery.xml`.  But, now that global "defaultSortOrder" is just the *first* `sortOption` combined with its `sortOrder`.  Is this accurate?

Overall, I agree with @artlowel that this is a bit confusing, as the `defaultSortOrder` settings in the `discovery.xml` really don't matter for most fields.  But, I don't have a better, immediate solution (without a larger overhaul of the `discovery.xml` config).

All in all, I think we all agree that this PR might be best to go forward as-is.  I just wanted to see if we first need to create any follow-up tickets for further cleanup.

**Comment by abollini** [comment]
hi,
the summary from @alemarte is accurate. The analysis of the angular behavior from @tdonohue also looks correct and reasonable to me.
On the REST side we now expose the configuration exactly as it is. Informing the client about which sort options exist and which is the preferred direction for each sort option.
How this information is used by the client, angular in our case, is mostly a decision of UX. For sure the client must use the first sort option and direction preference to inform the user about which is the current sorting when a search is performed without making explicit request about the sorting.
For the other sort options if as in our current angular application all the options / direction combinations are listed the preferred direction has limited or no value but in other UI where for instance the sort criteria is set clicking on a column it could allow to use the best direction for each criteria (still allowing the user to change direction if needed).
Think about a table layout where we show
score | title | date
when the user perform a default query we will get score **desc**
the user click on title, he want to get staff sorted by title **asc** but when he click on date usually he prefers to get the result sorted by date **desc**
this is what the REST configuration inform you.

We have explicitly avoid to constraint a sort field to a single direction because this was
- not supported on the REST side
- should require enforcement on the REST side to be well designed that in turn could limit administrative/unexpected use cases
- can be hard-coded on the angular side if really needed and related to UX as it was before (for the score sorting it was hardcoded to offer only a DESC option)

So my recommendation is to merge the current PRs as is or additionally hardcode the exclusion of the score ASC option only on the angular side (in this PR). If instead we really want to enforce/comunicate such constraint from the REST side a separate ticket is more appropriate

**Comment by tdonohue** [comment]
Thanks @abollini for the summary.  I agree that the REST API implementation does allow for more rich client implementations, even if our current client implementation is more limited in nature (but still a great start!).

As we have approvals on both frontend & backend PRs, I'm going to move forward with merging both.  It doesn't sound like we've identified any required follow-up work yet, but the discussion in this PR helps clarify the reason(s) for the change & future UI implementations we could someday consider.

**Pull request merged**
